### DISCUSSION TITLE
`Client._transport` may not exist during `Client.__del__()` or `Client.close()`

### DISCUSSION THREAD

**Discussion by j178** [encode/httpx#1540] [Potential Issue]
```py
>>> import httpx
>>> httpx.get("http://httpbin.org", proxies="socks5://127.0.0.1:1080")  # this will raise as we don't support socks proxy right now
>>> exit()
# then an `AttributeError` will occur here
```

![image](https://user-images.githubusercontent.com/10510431/112588683-c936ef00-8e3a-11eb-9e05-6c90298a340d.png)

https://github.com/encode/httpx/blob/c75ddc26c7e32e7c5fe99263899a187ff537aecc/httpx/_client.py#L609-L620

Since the error occured in `_get_proxy_map()`, before `_transport` was created, so `Client._transport` does not exist actuallly.

Then in `__del__` and `close()`, access to `_transport` will cause `AttributeError`:
https://github.com/encode/httpx/blob/c75ddc26c7e32e7c5fe99263899a187ff537aecc/httpx/_client.py#L1108-L1118

**Comment by lovelydinosaur** [comment]
Gotcha. Surprised me slightly that there was actually a client instance in scope there, that got garbage collected, because we never actually get to the point of returning one from `__init__()`.

Presumably the local exception context is hanging onto a reference to a partially instantiated client.

Pull request https://github.com/encode/httpx/pull/1541 neatens up this behaviour, and have confirmed in the console that it resolves this issue. I don't particularly want to futz around with a test case for it because it's  a bit of a weird fiddly edge case, and the change in behaviour feels self-evidently correct enough that I'm personally okay without one.
